CRM/Mailing/Service/ListUnsubscribe.php

```diff
@@ -0,0 +1,69 @@
+<?php
+
+/**
+ * Apply a full range of `List-Unsubscribe` header options.
+ *
+ * @service civi.mailing.listUnsubscribe
+ * @link https://datatracker.ietf.org/doc/html/rfc8058
+ */
+class CRM_Mailing_Service_ListUnsubscribe extends \Civi\Core\Service\AutoService implements \Symfony\Component\EventDispatcher\EventSubscriberInterface {
+
+  public static function getMethods(): array {
+    return [
+      'mailto' => ts('Mailto'),
+      'http' => ts('HTTP(S) Web-Form'),
+      'oneclick' => ts('HTTP(S) One-Click'),
+    ];
+  }
+
+  public static function getSubscribedEvents() {
+    return [
+      '&hook_civicrm_alterMailParams' => ['alterMailParams', 1000],
+    ];
+  }
+
+  /**
+   * @see \CRM_Utils_Hook::alterMailParams()
+   */
+  public function alterMailParams(&$params, $context = NULL): void {
+    // FIXME: Flexmailer (BasicHeaders) and BAO (getVerpAndUrlsAndHeaders) separately define
+    // `List-Unsubscribe: <mailto:....>`. And they have separate invocations of alterMailParams.
+    //
+    // This code is a little ugly because it anticipates serving both code-paths.
+    // But the BAO path should be properly killed. Doing so will allow you cleanup this code more.
+
+    if (!in_array($context, ['civimail', 'flexmailer'])) {
+      return;
+    }
+
+    $methods = Civi::settings()->get('civimail_unsubscribe_methods');
+    if ($methods === ['mailto']) {
+      return;
+    }
+
+    if (!preg_match(';^<mailto:u\.(\d+)\.(\d+)\.(\w*)@(.*)>$;', $params['List-Unsubscribe'], $m)) {
+      \Civi::log()->warning('Failed to set final value of List-Unsubscribe');
+      return;
+    }
+
+    $listUnsubscribe = [];
+    if (in_array('mailto', $methods)) {
+      $listUnsubscribe[] = $params['List-Unsubscribe'];
+    }
+    if (array_intersect(['http', 'oneclick'], $methods)) {
+      $listUnsubscribe[] = '<' . Civi::url('civicrm/mailing/unsubscribe')->addQuery([
+        'reset' => 1,
+        'jid' => $m[1],
+        'qid' => $m[2],
+        'h' => $m[3],
+      ]) . '>';
+    }
+
+    if (in_array('oneclick', $methods)) {
+      $params['headers']['List-Unsubscribe-Post'] = 'List-Unsubscribe=One-Click';
+    }
+    $params['headers']['List-Unsubscribe'] = implode(' ', $listUnsubscribe);
+    unset($params['List-Unsubscribe']);
```

Comment:
@artfulrobot I'm happy to go either way. The basic issue is that `$params` taxonomy is messy -- there are several ways to output those headers, and every option has a yucky angle:

```php
// (A)
$params['List-Unsubscribe'] = '<mailto:aaa@bbb.org> <https://ccc.org/unsubscribe/dddd>';
$params['headers']['List-Unsubscribe-Post'] = 'List-Unsubscribe=One-Click';
// Arg, that's asymmetric


// (B)
$params['headers']['List-Unsubscribe'] = '<mailto:aaa@bbb.org> <https://ccc.org/unsubscribe/dddd>';
$params['headers']['List-Unsubscribe-Post'] = 'List-Unsubscribe=One-Click';
// Arg, mailgunny needs to check both


// (C)
$params['List-Unsubscribe'] = '<mailto:aaa@bbb.org> <https://ccc.org/unsubscribe/dddd>';
$params['List-Unsubscribe-Post'] = 'List-Unsubscribe=One-Click';
// Arg, we're adding to the top-level mess
```

I was afraid that every option would be liable to break *some listener*.  The patch changes the data in `List-Unsubscribe` -- from one item to multiple items. Depending on how other listeners are written... and what they want to accomplish...  it might work cleanly... or blow up... or quietly do nothing. So if compatibility is murky no matter what, then at least we can look to accomplish "clean" -- and (B) is cleaner.

But OTOH, we do have `universe`. And `mailgunny` is the only listener in `universe` that touches `List-Unsubscribe` (*so it's the only listener that matters 😉 *). And the regex in `mailgunny` looks like it had enough foresight -- it probably is backward/forward/drop-in compatible with (A). And in that sense, compatibility is achievable.

So, yeah, I'll be OK with (A) or (B).

Reply:
We had a chat on CT call & got more votes for B